core/src/main/java/org/infinispan/configuration/cache/MemoryConfiguration.java

```diff
@@ -19,68 +20,160 @@
+    */
+   public boolean isOffHeap() {
+      return attributes.attribute(STORAGE).get() == StorageType.OFF_HEAP;
+   }
+
+   /**
+    * @return The max size in bytes or -1 if not configured.
+    */
+   public long maxSizeBytes() {
+      return maxSizeBytes;
+   }
+
+   public String maxSize() {
+      return attributes.attribute(MAX_SIZE).get();
+   }
+
+   public void maxSize(String maxSize) {
+      if (!isSizeBounded()) throw CONFIG.cannotIncreaseMaxSize();
+
+      maxSizeBytes = ByteQuantity.parse(maxSize);
+      attributes.attribute(MAX_SIZE).set(maxSize);
+   }
+
+   /**
+    * @return the max number of entries in memory or -1 if not configured.
+    */
+   public long maxCount() {
+      return attributes.attribute(MAX_COUNT).get();
+   }
+
+   public void maxCount(long maxCount) {
+      if (!isCountBounded()) throw CONFIG.cannotIncreaseMaxCount();
+
+      attributes.attribute(MAX_COUNT).set(maxCount);
    }
 
    /**
     * Storage type to use for the data container
     * @return
+    * @deprecated Use {@link #storage()} instead.
     */
+   @Deprecated
    public StorageType storageType() {
       return memoryStorageConfiguration.storageType();
    }
 
+   /**
+    * @return The memory {@link StorageType}.
+    */
+   public StorageType storage() {
+      return attributes.attribute(STORAGE).get();
+   }
+
    /**
     * Size of the eviction, -1 if disabled
-    * @return
+    * @deprecated Since 11.0, use {@link #maxCount()} or {@link #maxSize()} to obtain
+    * either the maximum number of entries or the maximum size of the data container.
     */
+   @Deprecated
    public long size() {
       return memoryStorageConfiguration.size();
    }
 
+   /**
+    * @deprecated Since 11.0, use {@link MemoryConfiguration#maxCount(long)} or
+    * {@link MemoryConfiguration#maxSize(String)} to dynamically configure the maximum number
+    * of entries or the maximum size of the data container.
+    */
+   @Deprecated
    public void size(long newSize) {
+      if (isSizeBounded()) {
+         attributes.attribute(MAX_SIZE).set(String.valueOf(newSize));
+      } else {
+         attributes.attribute(MAX_COUNT).set(newSize);
+      }
       memoryStorageConfiguration.size(newSize);
    }
 
    /**
     * The configured eviction type
-    * @return
+    * @deprecated Since 11.0, use {@link #maxCount()} or {@link #maxSize()} to obtain either the maximum number of
+    * entries or the maximum size of the data container.
     */
+   @Deprecated
    public EvictionType evictionType() {
       return memoryStorageConfiguration.evictionType();
    }
 
    /**
     * The configured eviction strategy
     * @return
+    * @deprecated Since 11.0, use {@link #whenFull()}
     */
+   @Deprecated
    public EvictionStrategy evictionStrategy() {
       return memoryStorageConfiguration.evictionStrategy();
    }
 
+   /**
+    * @return The configured {@link EvictionStrategy}.
+    */
+   public EvictionStrategy whenFull() {
```

Comment:
This isn't needed. We should be able to just maintain the eviction strategy in the MemoryConfiguration(Builder) directly and the MemoryStorageConfiguration(Builder) just invokes the parent when setting or retrieving the value.

I have a feeling we can do the same with size.

That is unless there was a reason we had to have separate config values for all of theses? But I would think that should reduce the complexity of the validate and other methods. This is what I did originally when changing eviction to memory element. https://github.com/infinispan/infinispan/blob/9.4.x/core/src/main/java/org/infinispan/configuration/cache/EvictionConfigurationBuilder.java#L82

You can also maintain the value in the child to see if it was set or not.

Reply:
This specific name ```when-full``` was a suggestion from @danberindei in the JIRA I had forgot to add. I think it was quite a expressive attribute name.

I separated on purpose the attributes to be able to detect when using legacy and new configs together, and throw an exception when they are mixed. 
The problem I found with syncing attributes one by one in the "setter" is that the JSON parser sets the ```Attribute``` values directly, and there's a ```read``` method that also sets attributes directly.
Some other places recycle builders, meaning they create a ```Configuration``` from the builder, then re-feed the ```Configuration``` into the builder's ```read``` and b then build again (one or more times). So yes, it's a bit complex but the config system is full of back doors...